File: examples/boards/raspberry-pi/src/bin/radio.rs
//! A LoRa radio on the header: an RFM95W breakout on SPI0, beaconing a reading and printing
//! every frame it hears in between.
//!
//! Wire the breakout's VIN to a 3V3 pin, GND to ground, SCK to GPIO11, MISO to GPIO9, MOSI to
//! GPIO10, CS to GPIO8 (CE0), and RST to GPIO25, and screw on an antenna for the band before
//! powering it. Turn SPI on, then run `cargo run --release --bin radio`. Two boards running it
//! hear each other. See docs/boards/raspberry-pi.md.

use std::error::Error;
use std::time::{Duration, Instant};

// ANCHOR: example
use pamoja_lora::budget::{Decibels, LinkBudget};
use pamoja_lora::region::Region;
use pamoja_radios::duty::DutyCycle;
use pamoja_radios::linux::{self, Wiring};
use pamoja_radios::radio::{RadioConfig, Reception};
use pamoja_radios::sx127x::config::PaOutput;
use pamoja_radios::sx127x::Board;

/// The header's first SPI chip select, the GPIO chip its lines are on, and the line the
/// breakout's reset pin is wired to.
const SPI: &str = "/dev/spidev0.0";
const CHIP: &str = "/dev/gpiochip0";
const RESET_LINE: u32 = 25;

/// The channel this node uses, the data rate it sends at, and how long it listens between
/// beacons.
const FREQUENCY_HZ: u32 = 868_100_000;
const DATA_RATE: u8 = 3;
const LISTEN: Duration = Duration::from_secs(10);

fn main() -> Result<(), Box<dyn Error>> {
    // The regional plan decides the channel's power ceiling and its duty cycle, so no limit
    // below is a number anyone has to remember.
    let plan = Region::Eu868.plan();
    let link = plan
        .link_settings(DATA_RATE)
        .expect("every LoRa data rate of the plan has link settings");
    let ceiling_dbm = plan.max_eirp_dbm(FREQUENCY_HZ);
    let permille = plan
        .duty_cycle_permille(FREQUENCY_HZ)
        .expect("the plan holds this channel");

    // A 2.15 dBi whip on half a decibel of pigtail. The antenna's gain counts against the
    // ceiling and the pigtail's loss counts for it, so the amplifier takes what is left.
    let whip = LinkBudget {
        transmit_antenna_gain_dbi: Decibels::from_hundredths(215),
        transmit_cable_loss_db: Decibels::from_tenths(5),
        ..LinkBudget::default()
    };
    let output_dbm = whip
        .max_transmit_power_dbm(Decibels::from_db(i32::from(ceiling_dbm)))
        .floor_db() as i8;

    // Opening resets the chip and reads its version back, so a wiring mistake is caught here
    // rather than on the first frame.
    let mut radio = linux::open_sx127x(
        &Wiring::new(SPI, CHIP, RESET_LINE),
        Board::new(PaOutput::PaBoost),
    )?;
    radio.configure(RadioConfig::new(FREQUENCY_HZ, link, output_dbm))?;
    println!("beacon on {FREQUENCY_HZ} Hz at DR{DATA_RATE}, {output_dbm} dBm under a {ceiling_dbm} dBm ceiling");

    // The duty cycle is the radio's other budget: each frame buys silence in proportion to its
    // airtime, and the guard says when the next one may go out.
    let mut duty = DutyCycle::new(permille);
    let clock = Instant::now();
    let mut buffer = [0u8; 255];
    let mut reading = 0u32;

    loop {
        // Listening returns as soon as a frame arrives, and a frame comes with the levels it
        // was heard at: how strong it was, and how far above the noise.
        match radio.receive(&mut buffer, LISTEN.as_micros() as u64)? {
            Reception::Frame { len, levels } => println!(
                "heard  {} at {:.0} dBm, SNR {:.1} dB",
                String::from_utf8_lossy(&buffer[..len]),
                decibels(levels.rssi_dbm),
                decibels(levels.snr_db)
            ),
            Reception::Corrupt => println!("heard  a frame whose CRC failed"),
            Reception::Timeout => {}
        }

        let now_us = clock.elapsed().as_micros() as u64;
        if duty.ready(now_us) {
            let frame = format!("pi reading {reading}");
            let airtime_us = radio.transmit(frame.as_bytes())?;
            duty.transmitted(now_us, &link, frame.len());
            println!("sent   {frame} in {airtime_us} us on air");
            reading += 1;
        }
    }
}

/// Returns a level in decibels, for printing.
fn decibels(value: Decibels) -> f64 {
    f64::from(value.hundredths()) / 100.0
}
// ANCHOR_END: example

File: bindings/dotnet/samples/Pamoja.Boards/RaspberryPi/Radio.cs
// ANCHOR: example
using System.Diagnostics;
using System.Text;

using Pamoja.Lora;
using Pamoja.Radios;

namespace Boards.RaspberryPi;

/// <summary>
/// A LoRa radio on the header: an RFM95W breakout on SPI0, beaconing a reading and printing
/// every frame it hears in between. Wire the breakout's VIN to a 3V3 pin, GND to ground,
/// SCK to GPIO11, MISO to GPIO9, MOSI to GPIO10, CS to GPIO8 (CE0), and RST to GPIO25, and
/// screw on an antenna for the band before powering it. Two boards running it hear each
/// other.
/// </summary>
public static class Radio
{
    // The header's first SPI chip select, the GPIO chip its lines are on, and the line the
    // breakout's reset pin is wired to.
    private const string Spi = "/dev/spidev0.0";
    private const string Chip = "/dev/gpiochip0";
    private const uint ResetLine = 25;

    // The channel this node uses, the data rate it sends at, and how long it listens
    // between beacons.
    private const uint FrequencyHz = 868_100_000;
    private const byte DataRate = 3;
    private static readonly TimeSpan Listen = TimeSpan.FromSeconds(10);

    /// <summary>Runs until the process is stopped.</summary>
    public static void Run()
    {
        // The regional plan decides the channel's power ceiling and its duty cycle, so no
        // limit below is a number anyone has to remember.
        using LoraChannelPlan plan = LoraChannelPlan.ForRegion(LoraRegion.Eu868);
        LoraLink link = plan.LinkSettings(DataRate)!;
        sbyte ceilingDbm = plan.MaxEirpDbm(FrequencyHz);
        uint permille = plan.DutyCyclePermille(FrequencyHz)!.Value;

        // A 2.15 dBi whip on half a decibel of pigtail. The antenna's gain counts against
        // the ceiling and the pigtail's loss counts for it, so the amplifier takes what is
        // left.
        var whip = new LoraLinkBudget { TransmitAntennaGainDbi = 2.15, TransmitCableLossDb = 0.5 };
        sbyte outputDbm = (sbyte)Math.Floor(whip.MaxTransmitPowerDbm(ceilingDbm));

        // Opening resets the chip and reads its version back, so a wiring mistake is caught
        // here rather than on the first frame.
        using LoraRadio radio = LoraRadio.OpenSx127x(
            new LoraRadioWiring(Spi, Chip, ResetLine), new Sx127xBoard(Sx127xPaOutput.PaBoost));
        radio.Configure(new LoraRadioConfig(FrequencyHz, link, outputDbm));
        Console.WriteLine(
            $"beacon on {FrequencyHz} Hz at DR{DataRate}, {outputDbm} dBm under a {ceilingDbm} dBm ceiling");

        // The duty cycle is the radio's other budget: each frame buys silence in proportion
        // to its airtime, and the guard says when the next one may go out.
        using var duty = new RadioDutyCycle(permille);
        var clock = Stopwatch.StartNew();
        int reading = 0;

        while (true)
        {
            // Listening returns as soon as a frame arrives, and a frame comes with the
            // levels it was heard at: how strong it was, and how far above the noise.
            LoraReception heard = radio.Receive(Listen);
            if (heard.Outcome == LoraReceptionOutcome.Frame)
            {
                Console.WriteLine(
                    $"heard  {Encoding.UTF8.GetString(heard.Payload!)} at {heard.RssiDbm:F0} dBm, SNR {heard.SnrDb:F1} dB");
            }
            else if (heard.Outcome == LoraReceptionOutcome.Corrupt)
            {
                Console.WriteLine("heard  a frame whose CRC failed");
            }

            ulong nowUs = (ulong)(clock.Elapsed.Ticks / (TimeSpan.TicksPerMillisecond / 1000));
            if (duty.Ready(nowUs))
            {
                string frame = $"pi reading {reading}";
                ulong airtimeUs = radio.Transmit(Encoding.UTF8.GetBytes(frame));
                duty.Transmitted(nowUs, link, frame.Length);
                Console.WriteLine($"sent   {frame} in {airtimeUs} us on air");
                reading++;
            }
        }
    }
}
// ANCHOR_END: example
